doc/decisions/0_drafts/errors_core.md

```diff
@@ -0,0 +1,222 @@
+
+This leaves the API design entirely open.
+
+TLS was only standardized in C11, while Elektra currently only requires C99.
+
+This has all the same issues as `errno`: needs to be cleared to detect errors, can only store one error (per thread), etc.
+
+### Pragmatic solution
+
+The pragmatic solution here is to simply assume "everything will be fine".
+
+Many functions only return errors for simple precondition violations, like passing a `NULL` pointer or otherwise invalid argument.
+Especially for `libelektra-core` (the only place where `Key * errorKey` truly cannot be used), all functions must be side effect free.
+Therefore, the only possible cause of errors are invalid arguments.
+
+If we now assume the function is written such that it fails before it changes any argument's data, then we can infer that only a single error indicating return value is needed.
+Because the arguments are unchanged and the only possible cause of errors, we can distinguish between errors after the fact.
+
+As stated, many functions' only error cases are precondition violations.
+If the preconditions are simple enough, they can be checked beforehand if it is not clear whether the precondition holds.
+
+Take for example some hypothetical `keyValue` function, in a world where we have a `key->value` container for COW that should always be present.
+Looking just at the error checks, without even considering the function signature:
+
+```c
+if (key == NULL) {
+// error 1: precondition violation
+}
+
+if (key->value == NULL) {
+// error 2: invalid Key, COW container missing
+}
+
+// success: return value
+return key->value->data;
+```
+
+If we want separate error indications for both error cases and a nice API we immediately hit a problem.
+The obvious return type is `const void *`, because that's the type of the key value (and we want it const).
+But that's a pointer, so only `NULL` could be used for errors.
+The issue is, `NULL` may be a valid key value.
+So what do we do?
+
+Essentially, we do nothing, "everything is fine":
+
+```c
+const void * keyValue (Key * key) {
+  if (key == NULL) {
+    return NULL;
+  }
+
+  assert(key->value != NULL);
+  if (key->value == NULL) {
+    return NULL;
+  }
+
+  return key->value->data;
+}
+```
+
+We return `NULL` for both error cases and even let the errors overlap with a legitimate success value.
+In the first case, it's fine for multiple reasons:
+
+- Many callers won't care, if the value inside the key is `NULL` or the key itself is `NULL`.
+  For example, `keyValue (ksLookup (ks, name)) == NULL` is probably fine either way.
+- If the caller does care, they can easily check `if (key == NULL)` before _or_ after the `keyValue` call.
+
+The second case has a different reason.
+The error is not the caller's fault.
+It is an internal issue the caller can do nothing about, `key->value` should always be there.
+That's why we add an `assert` before the check.
+This will trigger in a debug build, where the internal error can be fixed.
```

Comment:
I'm not sure what you mean here....

Yes, that's the point. Only Elektra's debug builds should check these internal conditions. The debug builds are used in our CI (at least I hope so) to run all our tests. That should guard against these issues.

In a release build I think it is bad form for a library to call `abort()` and kill the process. Especially, since we definitely could do something else (like `return NULL`) in this case.

A further note (which I probably should add) the extra `if` check after the `assert` should only be there, if it prevents something like a segfault (like in this case). If it's something more benign like "if the assert doesn't hold, we leak memory" then there shouldn't be an extra check.

However, this is the part of the suggested solution I'm most flexible about. I do think we should have separate rules for internal conditions that the user may not even be able to check. But I'm open to different solutions for them.